프로젝트 관리 앱 [STEP 2] 현이

seohyeon2 opened the pull request:
안녕하세요 제이슨! (@ehgud0670) ✨
미해결된 부분을 계속 붙잡고 있다가 시간이 벌써 이만큼 지났네요..🥲
끝까지 해결해보고 싶었는데 시간이 더 지체될 것 같아 먼저 PR 보냅니다..!

미숙한 부분이 많은 코드지만 이번 스텝도 잘 부탁드립니다!☺️

---

<br>

## 🥸 코드 설명

- 폴더 구조
    - 할일을 생성, 수정, 조회 하는 뷰(`TodoFormViewController`)는 **MVP**로 구성하려고 했습니다.
    - 메인 뷰(`MainHomeViewController`)는 **MVVM**으로 구성하려고 했습니다.
    - 구조를 나누다 보니 Database 관련 파일이나, Model, View는 겹치는 부분이 있어 `Common`으로 빼주었습니다!

```swift
.
└── ProjectManager/
    ├── ProjectManager
    ├── info.plist
    ├── Common/
    │   ├── Database
    │   │   ├── DatabaseManager
    │   │   └── DRealmDatabaseManager
    │   ├── Model
    │   │   ├── RealmDatabaseModel
    │   │   ├── TaskModel
    │   │   └── TaskData
    │   ├── View
    │   │   ├── Main.storyborad
    │   │   └── TableViewCell
    ├── MainHome/
    │    ├── ViewModel/
    │    │   └── MainHomeViewModel
    │    └── MainHomeViewController
    └── TodoForm/
        └── Controller/
            └── TodoFormViewController
```

- 구조 적용
    - `TodoFormViewController` 에서 할일을 새로 만들 때, 수정할 때 모델에 변화가 생긴다고 생각하여 이 부분을 `Notification`을 사용하여 처리했습니다.
    - 제가 생각한 방향이 MVP를 잘 지키지 못했다면 가감없이 피드백 해주시면 감사하겠습니다 🤩
    - `MainHomeViewController` 에서 바인딩을 사용해보려고 했지만, 제대로 적용을 하지 못했습니다..😢 


<br>


## ⭐️ 궁금한 점 및 조언 얻고 싶은 점


### 1. `popOver` 위치
- 할 일(cell)을 누를 때, 누른 곳에 팝오버가 보이게 코드를 작성했습니다.
- `Todo`의 테이블뷰에서는 잘 동작하는데, `Doing`과 `Done`에서는 팝오버가 제대로 나타나지 않습니다🥲
- 현재는 아래의 코드로 사용하고 있는데, **`location` 대신 cell을 클릭한 위치에서 팝오버가 뜨게끔 할 수 있는 방법이 있을까요? 😯**
- 참고로 이 부분이 제대로 해결되지 못해 상황에 따라 메뉴를 나누지 못했습니다..😭 (빠르게 해결해보도록 하겠습니다..!!)
```swift
 @objc func handleLongPressGesture(recognizer: UITapGestureRecognizer) {
     let location = recognizer.location(in: recognizer.view)
     
     // ... 코드 생략
     
     popover?.sourceRect = CGRect(x: location.x, y: location.y + 60 , width: 
}
``` 

### 2. UIButton 바인딩
- `MainHomeViewController`에 있는 프로퍼티 중 UIButton은 Todo, Doing, Done이 몇개 있는지 알려주는 역할을 합니다.
- 터치 이벤트를 위해 만든게 아닌, 백그라운드에 이미지를 넣고자 만든 프로퍼티입니다.
- 처음엔 label에 백그라운드 이미지를 주려고 했지만 결국 실패하여 아래와 같이 `Backgroud Configuration > Image` 를 사용하였습니다..😅
<img src = "https://i.imgur.com/RmDZFPl.png" width = "400" height = "250">

- **Todo, Doing, Done의 개수가 바뀔 때마다 Button의 title을 변경하고자 하는데, 이럴 경우 어떻게 바인딩을 진행하면 좋은지 궁금합니다..!**
- 또 이 상태에서 바인딩을 진행하는 것보다 label에 백그라운드 이미지를 줄 수 있는 코드를 작성하여 바인딩을 하는 것이 더 나은 방법인지도 궁금합니다😌

seohyeon2 committed:
apply: SwiftLint 적용

seohyeon2 committed:
feat: 가로모드만 지원되도록 구현

seohyeon2 committed:
design: storyboard를 사용하여 홈 화면 구성함

seohyeon2 committed:
create: TableViewCell 파일 생성

seohyeon2 committed:
design: tableViewCell 오토레이아웃 적용

seohyeon2 committed:
apply: RealmSwift 적용

seohyeon2 committed:
create: Local Database에 필요한 파일 생성
- DatabaseManager 파일 생성
- RealmDatabaseManager 파일 생성
- RealmDatabaseModel 파일 생성
- TaskModel 파일 생성

seohyeon2 committed:
feat: realmDatabase의 create 및 read 메서드 구현

seohyeon2 committed:
feat: realmDatabase의 delete 메서드 구현

seohyeon2 committed:
feat: realmDatabase의 update 메서드 구현

- DatabaseManager의 updateDatabase 메서드 매개변수 수정
- RealmDatabaseManager의 readDatabase 메서드 수정
- RealmDatabaseModel에 업무 상태와 아이디 추가
- TaskModel에 업무 상태와 아이디 추가
- TaskState 타입 정의

seohyeon2 committed:
feat: 업무 상태별 개수를 구해주는 getTaskStateCount 메서드 구현

seohyeon2 committed:
refactor: tableView의 cell을 seedData에서 Database 내용으로 변경

- todoList, doingLis, doneList 배열 생성
-  setUpDataList 메서드 구현

seohyeon2 committed:
design: cell 제목, 설명, 기한에 대한 표시 수정

seohyeon2 committed:
design: 할일 추가 및 수정 화면 구성

seohyeon2 committed:
create: 할일 추가 및 수정 할 수 있는 TodoFormViewController 파일 생성

- TextField, TextView 그림자 적용

seohyeon2 committed:
feat: 팝오버 기능 구현

- 길게 터치하면 팝오버 뜨는 기능 구현
- 길게 터치하는 제스처와 스크롤 동시에 가능하게 구현
- 팝오버를 통해 업무 상태 변경되게 구현

seohyeon2 committed:
feat: 왼쪽으로 스와이프하여 삭제하는 기능 추가

seohyeon2 committed:
feat: 할일 추가하는 기능 구현

- didTapCancelButton 메서드 구현
- didTapDoneButton 메서드 구현

seohyeon2 committed:
chore: 프로젝트 폴더 정리

seohyeon2 committed:
create: 업무 관련 데이터를 처리하는 TaskData 파일 생성

- DatabaseManager 싱글톤으로 구현

seohyeon2 committed:
create: MainHomeViewModel 파일 생성

seohyeon2 committed:
refactor: MainHomeViewController에서 팝오버로 상태 옮기는 로직 수정

seohyeon2 committed:
refactor: MainHomeViewController의 viewDidLoad 메서드 코드 정리

- setUpGestureEvent  메서드로 분리
- setUpListCount 메서드로 분리
- setUpTableViewDelegate 메서드로 분리
- setUpTableViewDataSource 메서드로 분리

seohyeon2 committed:
feat: 할일 목록 데이터 반환해주는 getDataList 메서드 구현

- fetchDataList 메서드 구현
- MainHomeViewController에서 불필요한 프로퍼티 및 코드 삭제
-  getDataList 메서드 구현에 따른 로직 변경

seohyeon2 committed:
refactor: 할일 추가 시, Notification으로 알림오게 수정

seohyeon2 committed:
feat: 새롭게 생성된 Todo를 추가하는 addTodo 메서드 구현

- 오타 및 코딩컨벤션 수정
- 테이블 뷰 reload 로직 수정

seohyeon2 committed:
refactor: 데이터 삭제 로직 수정

- RealmDatabaseManager에  deleteDatabase, updateDatabase 메서드 수정
-  DatabaseManager 프로토콜 메서드 매개변수 수정
-  RealmDatabaseManager에 search 메서드 구현
- MainHomeViewModel에 remove 메서드 구현

seohyeon2 committed:
chore: 폴더 및 컨벤션 정리

seohyeon2 committed:
feat: 할일 저장 시, 원하는 형태로 저장해주는 getCurrentDateTime 메서드 구현

seohyeon2 committed:
feat: 기한 지난 날짜 검증하는 checkPastDate 메서드 구현

seohyeon2 committed:
feat: cell 클릭 시 상세화면 볼 수 있는 setUpForm 메서드 구현

- 스토리보드에서 segue 연결
- 노티피케이션 post, add 추가
- setUpGestureEvent 메서드 수정
- 데이터 하나 전달해주는 readData 메서드 구현

seohyeon2 committed:
refactor: cell 선택 시, 할일 업데이트 가능하게 수정

seohyeon2 committed:
feat: 앱 시작 시, 로컬데이터베이스 초기화하는  deleteAllDatabase 메서드 구현

seohyeon2 committed:
chore: 코딩컨벤션 수정

seohyeon2 committed:
refactor: 팝오버에 TODO로 이동하는 버튼 추가

ehgud0670 commented on ProjectManager/ProjectManager/Common/AppDelegate.swift:
```diff
@@ -0,0 +1,39 @@
+        _ application: UIApplication,
+        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
+    ) -> Bool {
+        RealmDatabaseManager().deleteAllDatabase()
```
RealmDatabaseManager를 새로 생성해도 기존 데이터베이스에 있는 데이터를 제거 할 수 있다면 
RealmDatabaseManager도 shared (싱글톤) 객체를 생성해주는 게 낫지 않을까요?!
지금은 RealmDatabaseManager 여러개 생성해도 구분이 없어보입니다 🧐

seohyeon2 replied:
TaskDataManager(기존 TaskData)의 싱글톤으로 교체했습니다!

seohyeon2 replied:
처음엔 싱글톤으로 구현했지만,
프로젝트 관리앱2에서 해당 코드는 사용하지 않게 되었습니다.

`RealmDatabaseManager`는 `MainHomeViewModel`에서만 사용하여 
최종적으로 싱글톤 삭제하고 `MainHomeViewModel`에서 프로퍼티로 갖고 있게 수정했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/TodoForm/Controller/TodoFormViewController.swift:
```diff
@@ -0,0 +1,113 @@
+            state: TaskState.todo
+        )
+
+        TaskData.shared.databaseManager.createDatabase(data: data)
```
RealmDatabaseModel을 그대로 생성해서 전해주는 것보다 TaskModel을 생성해서 건네주는 건 어떨까요? 
Realm을 안쓰고 다른 데이터베이스로 교체한다면 위 코드는 수정해야 해서요.

seohyeon2 replied:
`create` 메서드의 반환 값을 `TaskModel` 로 수정했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/Common/Database/RealmDatabaseManager.swift:
```diff
@@ -0,0 +1,94 @@
+        guard let realm = realm,
+              ((try? realm.write({ realm.add(data) })) != nil) else {
+            return
+        }
```
여기도 deleteDatabase 메소드 처럼 write 실패시에 print 라도 해서 오류가 발생했다 것을 표시하면 더 좋을 것 같아요 :)

seohyeon2 replied:
일단 `print`로 에러처리 했습니다!
추후 에러를 알림창으로 보여주게 수정하겠습니다😌

ehgud0670 commented on ProjectManager/ProjectManager/Common/Database/RealmDatabaseManager.swift:
```diff
@@ -0,0 +1,94 @@
+        }
+    }
+
+    func readDatabase() -> [TaskModel] {
```
```suggestion
    func read() -> [TaskModel] {
```
객체 이름자체가 Database 가 포함되어 있으니 중복으로 보입니다. 굳이 없어도 될 것 같아요 :)

seohyeon2 replied:
삭제해주었습니다😁

ehgud0670 commented on ProjectManager/ProjectManager/Common/Database/RealmDatabaseManager.swift:
```diff
@@ -0,0 +1,94 @@
+            return [TaskModel]()
+        }
+
+        var database = [TaskModel]()
```
```suggestion
        var taskModels = [TaskModel]()
```

taskModels라고 하는게 더 나아보이네요 :) 
데이터베이스는 아니니까요!

seohyeon2 replied:
네이밍 수정 했습니다😄

ehgud0670 commented on ProjectManager/ProjectManager/Common/Database/RealmDatabaseManager.swift:
```diff
@@ -0,0 +1,94 @@
+        var database = [TaskModel]()
+        let realmDatabase = realm.objects(RealmDatabaseModel.self)
+
+        realmDatabase.forEach { realmDatabaseModel in
```
forEach 보다는 Map 을 써보는건 어떨까요?

seohyeon2 replied:
`Map`으로 변경했습니다😃

ehgud0670 commented on ProjectManager/ProjectManager/Common/Database/RealmDatabaseManager.swift:
```diff
@@ -0,0 +1,94 @@
+            taskModel.taskDescription = realmDatabaseModel.taskDescription
+            taskModel.taskDeadline = realmDatabaseModel.taskDeadline
+            taskModel.taskState = realmDatabaseModel.taskState
+            taskModel.id = realmDatabaseModel.id
```
TaskModel을 생성해서 프로퍼티 세팅하는것보단 init 에서 해주는게 할당하는 프로퍼티를 빼먹지 않을 수 있다는 점에서 더 좋아 보입니다 :)

seohyeon2 replied:
`init`으로 수정했습니다🙂

ehgud0670 commented on ProjectManager/ProjectManager/Common/Model/TaskData.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+class TaskData {
```
TaskData가 필요할 이유가 있을까요? 이거 없이 RealmDatabaseManager가 싱글톤 객체를 가지면 될 것 같아서요.
그리고 그냥 모델값처럼 보여서 네이밍 변경이 필요해 보입니다(ex. `TaskDataManager`).

seohyeon2 replied:
회의실에서 말씀해주신 것처럼 `ViewModel`이 `DataBaseProtocol` 타입의 프로퍼티를 갖게 수정하고, `TaskData`도 삭제했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/Common/Model/TaskModel.swift:
```diff
@@ -0,0 +1,22 @@
+    var id: UUID?
+}
+
+enum TaskState {
+    static let todo = "TODO"
+    static let doing = "DOING"
+    static let done = "DONE"
+}
```
이건 상수로 말고 case 로 구분하는게 더 자연스러워 보이는데 어떠신가요? 
title 이 필요하면 연산프로퍼티로 따로 빼면 되고요.

seohyeon2 replied:
case로 변경 후, 연관값 사용하는 것으로 수정했습니다😎

ehgud0670 commented on ProjectManager/ProjectManager/Common/View/TableViewCell.swift:
```diff
@@ -0,0 +1,45 @@
+        deadlineLabel.text = data.taskDeadline
+
+        if checkPastDate(date: data.taskDeadline) {
+            deadlineLabel.textColor = .red
```
checkPastDate 메소드를 TaskModel  로 빼서 아래처럼 만들어보는건 어떨까요?

```suggestion
            deadlineLabel.textColor = data.deadLineLabelTextColor
```

seohyeon2 replied:
제안해주신 코드로 수정했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,190 @@
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        selectedIndex = indexPath.row
+        setUpTaskState(tableView: tableView)
+        setUpGestureEvent(tableView)
```
gesture는 셀렉트할때마다 설정해주는 것이 아니라 viewDidLoad 에서 한번만 설정해주면 되지 않나요?? 🧐

seohyeon2 replied:
tableView 매개변수를 사용하려고 셀렉트 할 때마다 호출해주었습니다😅
로직 수정으로 더이상 `tableView`가 필요하지 않아,  `viewDidLoad`에서 한 번 호출되게 수정했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        selectedIndex = indexPath.row
```
selectedIndex라는 값도 ViewModel에 위치해주는 게 자연스러워 보입니다.

seohyeon2 replied:
로직 변경으로 `selectedIndex`가 필요하지 않아 최종 삭제했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,190 @@
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+
+        let cell = todoTableView.dequeueReusableCell(withIdentifier: "todoTableViewCell", for: indexPath) as! TableViewCell
```
guard let 을 쓰는게 더 좋아보입니다 :)

seohyeon2 replied:
앗 바로 수정했습니다!!

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,190 @@
+
+extension MainHomeViewController: UITableViewDelegate, UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        if tableView == todoTableView {
```
`==` 보다 `===` 를 쓰는게 더 옳은 것 같은데 어떠신가요? 🧐

seohyeon2 replied:
`===`를 쓰면 동일한 인스턴스를 참조하고 있는지 알 수 있기 때문에,  현재 상황에서는 `===` 쓰는  게옳은 것 같습니다☺️

`===`로 수정했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/TodoForm/Controller/TodoFormViewController.swift:
```diff
@@ -0,0 +1,113 @@
+            state: TaskState.todo
+        )
+
+        TaskData.shared.databaseManager.createDatabase(data: data)
+        notifyChangedModel()
+    }
+
+    private func notifyChangedModel() {
+        NotificationCenter.default.post(name: NSNotification.Name("모델 리로드 예정"), object: nil)
+        dismiss(animated: true)
+    }
```
dismiss 할때 노티로 homeVC에 알리기 보다는 delegate 패턴을 통해 알리는 것이 어떨까요? 
그리고 TodoFormViewController 에서 직접 데이터베이스에 모델값을 create 하기 보다는 일단은 델리게이트 패턴을 통해 
homeVC에 모델값을 전달한 후 homeVC의 viewModel이 데이터베이스에 모델값을 추가하는 것이 괜찮아보이는 데 어떠신지.. 🧐

seohyeon2 replied:
TodoFormViewController에서는 delegate 패턴으로 ViewModel에게 데이터 전달만 하도록 수정했습니다😁

seohyeon2 committed:
docs: Podfile 추가

ehgud0670 reviewed (commented):
빌드 해보고 더 리뷰 드릴 수 있으면 드려볼게요! 
수고 하셨습니다 :)

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,190 @@
+
+        let popover = actionSheet.popoverPresentationController
+        popover?.sourceView = view
+        popover?.sourceRect = CGRect(x: location.x, y: location.y + 120 , width: 60, height: 60)
```
> 궁금한 점 

x, y 적용할 때 recognizer.view.frame 의 x,y값도 고려해주면 popOver 의 위치가 잘 나오지 않을까요? 🧐

seohyeon2 replied:
말씀해주신 `recognizer.view.frame`을 사용하여 수정했습니다😌

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/ViewModel/MainHomeViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+        doingList = [TaskModel]()
+        doneList = [TaskModel]()
+
+        allData.forEach { data in
```
forEach 를 쓰기보다는 filter를 이용해 todoList, doingList, doneList 배열 값을 갱신하는 건 어떨까요?

seohyeon2 replied:
`filter` 로 수정했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/ViewModel/MainHomeViewModel.swift:
```diff
@@ -0,0 +1,87 @@
+        doneList = [TaskModel]()
+
+        allData.forEach { data in
+            currentState = data.taskState
```
currentState라는 프로퍼티가 있는것보단 없는게 나아보입니다 🧐

seohyeon2 replied:
`currentState` 프로퍼티 대신 함수의 인자값으로 받아오게 수정했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,190 @@
+    @IBOutlet weak var todoTableView: UITableView!
+    @IBOutlet weak var doneTableView: UITableView!
+
+    @IBOutlet weak var todoCount: UIButton!
```
> 궁금한 점 2

저는 UILabel을 만들고 cornerRadius를 통해 원이 되게 만들듯 합니다 :)

seohyeon2 replied:
아이디어 감사합니다! `cornerRadius` 이용하여 수정했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    private func reloadTableView() {
+        todoTableView.reloadData()
+        doingTableView.reloadData()
+        doneTableView.reloadData()
+
+        setUpTableViewCount()
```
> 궁금한 점 3

저도 노티를 이용해 tableview의 count를 갱신할 것 같아요 :)

seohyeon2 replied:
클로저를 이용해보고 싶어서 노티대신 클로저로 갱신되게 변경 했습니다😌

ehgud0670 reviewed (commented):
고생하셨습니다 :)

ehgud0670 commented:
https://user-images.githubusercontent.com/38216027/190321398-2e192864-caa3-4a53-8571-d5b8c1fd7c5a.mov
위 영상을 보시면 잘못된 데이터를 옮기는 오류가 있습니다. 3을 옮기려는데 2을 옮김

해결책: tableView 말고, cell에 LongPressGesture를 add 하고 cell의 Delegate 를 만들어 처리하게 하는 것도 좋은 방법일 수 있을 것 같아요 🧐

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,190 @@
+        setUpGestureEvent(tableView)
+
+        NotificationCenter.default.post(name: NSNotification.Name("모델 수정"), object: viewModel.readData(index: indexPath.row))
+        tableView.deselectRow(at: indexPath, animated: false)
```
Cell의 selctionStyle을 none으로 설정하는 건 어떨까요?!

seohyeon2 replied:
Cell의 `selctionStyle`을 `none`으로 할 경우, 
cell의 backgroundColor가 누른 전후로 동일하여
어떤 cell을 선택했는지, 선택이 되긴 했는지 의문점이 들 수 있겠다 생각했습니다.

사용자로 하게끔 선택이 잘 됐는지 보여주려고 일부로 추가한 부분인데,
`tableView.deselectRow(at: indexPath, animated: false)` 코드가 없어도
UX적으로 괜찮을까요?😲

제이슨은 어떻게 생각하시는지 궁금합니다!☺️

ehgud0670 replied:
그런 의도가 있으셨군요! 좋습니다 :)

ehgud0670 commented:
노티 말고 클로저를 통한 바인딩 방식이 있음을 알려드립니다 :) 
아래 링크가 예시입니다.

https://gist.github.com/godrm/6914218e567de542bb653b0c4b99f1da

seohyeon2 committed:
refactor: Database관련 타입 네이밍 수정

- AppDelegate안의 데이터 지워주는 코드를 싱글톤으로 수정
- DatabaseManager에서 DatabaseProtocol로 네이밍 수정
- RealmDatabaseManager에서 RealmDatabase로 네이밍 수정
- TaskData에서 TaskDataManager로 네이밍 수정

seohyeon2 committed:
refactor: RealmDatabase 파일 수정

- create 메서드 매개변수 타입을 Object에서 TaskModel로 변경
- 모든 메서드명에서 Database 삭제
- create, deleteAll 메서드 do-catch문 삽입
- read 메서드 안에 forEach문을 Map으로 변경
- read 메서드 안에서 사용하는 프로퍼티 네이밍 수정 및 TaskModel 생성 시 초기화로 데이터 넣어줌

seohyeon2 committed:
refactor: checkPastDate메서드 위치 변경

- TableViewCell에서 TaskModel로 변경함

seohyeon2 committed:
refactor: 데이터 전달 방식을 Notification에서 delegate로 수정

- storyboard 네임스페이스 처리를 위한 ReuseIdentifying 프로토콜 생성
- 데이터 전달을 위한 SendDelegate 프로토콜 생성
- 스토리보드 segment 연결 해제 후, 코드로 화면이동 구현

seohyeon2 committed:
refactor: notifyChangedModel 메서드 변경

- Notification에서 delegate로 수정

seohyeon2 committed:
design: 할일별 개수 알려주는 button을 label로 변경

seohyeon2 committed:
refactor: 할일별 개수 나타내는 메서드를 binding 방식으로 변경

seohyeon2 committed:
create: TaskState 파일 생성

seohyeon2 committed:
refactor: MainHomeViewController 및 TaskState 로직 수정

- TaskState의 static let을 case, 연관값으로 수정
- MainHomeViewController의 setUpTableViewDataSource(), setUpTableViewDelegate(), reloadTableView() 고차함수로 로직변경

seohyeon2 committed:
refactor: TableView와 관련된 메서드 수정

- TableViewCell storyboard 네이밍을 todoTableViewCell에서 TableViewCell으로 변경
- TableViewCell 타입 ReuseIdentifying 채택
- numberOfRowsInSection 메서드 로직 및 컨벤션 수정
- remove 메서드 호출 시, fetchDataList 메서드 호출하게 수정

seohyeon2 committed:
refactor: storyboard에서 segue로 화면 이동한 것을 action 방식으로 변경

- didTapAddButton 메서드 생성
- TodoFormViewController 화면 보여주는 presentTodoForm 메서드 생성
- TodoFormViewController에 delegate 프로퍼티 생성하여 로직 수정

seohyeon2 committed:
refactor: MainHomeViewModel의 fetchDataList 로직 변경

- forEach 메서드 이용에서 filter 메서드 이용으로 변경

seohyeon2 committed:
refactor: MainHomeViewModel의 currentList 프로퍼티 삭제에 따른 코드 수정

seohyeon2 committed:
refactor: popover 관련 코드 수정

- TaskState 타입 CaseIterable 채택, title 및 other 프로퍼티 생성
- setUpGestureEvent 메서드 호출 위치 viewDidLoad로 변경
- setUpGestureEvent 메서드에 tap 이벤트 추가
- [UIAlertAction] 만들어주는 getActionButton 메서드 구현
- 선택된 cell이 어떤 상태인지 알려주는 getTaskState 메서드 구현
- handleLongPressGesture 메서드 구조 변경

seohyeon2 committed:
chore: 코드 정리 및 ViewController에 MARK 주석 추가

seohyeon2 committed:
refactor: TapGestureRecognizer 삭제

seohyeon2 committed:
refactor: TaskModel에 deadLineLabelTextColor 프로퍼티 추가 및 코드 정리

seohyeon2 committed:
refactor: setUpCell 메서드 수정

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/ViewModel/Observable.swift:
```diff
@@ -0,0 +1,23 @@
+//  Created by seohyeon park on 2022/09/17.
+//
+
+class Observable<T> {
+  typealias Listener = (T) -> Void
+  var listener: Listener?
+  var value: T {
+    didSet {
+      listener?(value)
+    }
+  }
+  init(_ value: T) {
+    self.value = value
+  }
+  func bind(listener: Listener?) {
+    self.listener = listener
+    listener?(value)
+  }
+}
```
```suggestion
class Observable<T> {
  typealias Listener = (T) -> Void
  var listener: Listener?
  
  var value: T {
    didSet {
      listener?(value)
    }
  }
  
  init(_ value: T) {
    self.value = value
  }
  
  func bind(listener: Listener?) {
    self.listener = listener
    listener?(value)
  }
}
```
=> 함수 마다 한칸씩은 띄워주시는게 가독성면에서 좋습니다 :)

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,297 @@
+            self?.doneCountLabel.text = count.description
+        }
+
+        reloadTableView()
```
* 바인드 코드에 reloadTableView()는 관련이 없습니다. 분리해야 할 듯 합니다. 🧐
* 아니면 reloadTableView()도 bind 함수의 클로저 코드안에 있어야 해요.
  * 현재 구현되어있는 bind 코드 말고 테이블뷰 데이터 관련되어 bind 코드를 따로 만들라는 말씀입니다 :)

seohyeon2 replied:
분리했습니다!

ehgud0670 commented on ProjectManager/ProjectManager/MainHome/MainHomeViewController.swift:
```diff
@@ -0,0 +1,297 @@
+        }
+
+        viewModel.changeList(data: data)
+        bind()
```
bind는 viewDidLoad에 한번만 세팅해주면 될 듯 합니다. 
viewDidLoad에서 bind 한번만 호출시키고 여기서는 chageList 메소드를 호출해주면 알아서 뷰가 업데이팅 되도록 코드가 수정되어야 할 것 같아요.

seohyeon2 replied:
bind는 viewDidLoad에서 한 번만 호출되게 수정했습니다!

ehgud0670 reviewed (commented):
잘해주고 계시네요 :) 
고생하셨습니다. 
추가 피드백 확인해주세요 💪

seohyeon2 committed:
chore: 코드 정리

Co-authored-by: Doe-Hyung KIM <kimdo2297@gmail.com>

seohyeon2 committed:
refactor: Observable 삭제에 따른 코드 수정
